## Packages used for the project

In [1]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer
import pandas as pd
import requests
import os
%matplotlib inline
#import matplotlib
import matplotlib.pyplot as plt
import numpy as np


## Gather

#### Saving the files provided by Udacity in the workbook:

In [2]:
## This where gather the three required data file 
# WeRateDogs Twitter archive provided by the instructor:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

# The tweet image predictions by the instructor:
urlLink = " https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(urlLink)
open('image_predictions.tsv', 'wb').write(response.content)

image_predictions = pd.read_csv('image_predictions.tsv', sep= '\t' )


#### Gather other tweets using tweepy API and save it in tweet_json.txt

In [3]:
### Below function will connect to tweepy API and return tweepy API
def conn_tweep_api() :
    consumer_key = '111111'
    consumer_secret = '222'
    access_token = '3-33424'
    access_secret = '23424234'

    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_secret)

    api = tweepy.API(auth, wait_on_rate_limit_notify=True, wait_on_rate_limit=True)
    
    return api


#### Please note that I have tested below code and it is working properly without any issue. However, for the sake of avoiding the Jupyter book crashs (it happens to me many time), I used the tweet_json.txt provided by Udacity.

In [4]:
# To avoid running the code each time I run all the blocks, I have the
# the following if condition to check if the file existed or not
# if the file exist then there no need get the tweets again

if not (os.path.isfile('./tweet_json.txt')) :
    api = conn_tweep_api()
    print ('Test')
    
    with open('tweet-json.txt', 'a' ,encoding='utf8') as f :
        for tweet_id in twitter_archive['tweet_id']:
            try:
                tweet =api.get_status(tweet_id, tweet_mode='extended')
                json.dump(tweet._json, f)
                f.write('\n')
        
            except tweepy.TweepError as e:
                print (e)
                print ('Fail ', tweet_id)

#### Create a DataFrame from the tweet_json.txt file
 reference (https://stackoverflow.com/questions/47925828/how-to-create-a-pandas-dataframe-using-tweepy)

In [5]:
# read the JSON into pandas data frame and set the index
tweet_json = pd.read_json('tweet_json.txt', lines=True)
tweet_json = tweet_json[['id','retweet_count','favorite_count',  'retweeted_status']]


## Assess

#### Verify that I have gathered all the files correctly:

In [6]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [7]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [8]:
tweet_json

id  retweet_count  favorite_count retweeted_status
0     892420643555336193           8853           39467              NaN
1     892177421306343426           6514           33819              NaN
2     891815181378084864           4328           25461              NaN
3     891689557279858688           8964           42908              NaN
4     891327558926688256           9774           41048              NaN
5     891087950875897856           3261           20562              NaN
6     890971913173991426           2158           12041              NaN
7     890729181411237888          16716           56848              NaN
8     890609185150312448           4429           28226              NaN
9     890240255349198849           7711           32467              NaN
10    890006608113172480           7624           31166              NaN
11    889880896479866881           5156           28268              NaN
12    889665388333682689           8538           38818              NaN
13    889638837579907072           4735           27672              NaN
14    889531135344209921           2321           15359              NaN
15    889278841981685760           5637           25652              NaN
16    888917238123831296           4709           29611              NaN
17    888804989199671297           4559           26080              NaN
18    888554962724278272           3732           20290              NaN
19    888078434458587136           3653           22201              NaN
20    887705289381826560           5609           30779              NaN
21    887517139158093824          12082           46959              NaN
22    887473957103951883          18781           69871              NaN
23    887343217045368832          10737           34222              NaN
24    887101392804085760           6167           31061              NaN
25    886983233522544640           8084           35859              NaN
26    886736880519319552           3443           12306              NaN
27    886680336477933568           4610           22798              NaN
28    886366144734445568           3316           21524              NaN
29    886267009285017600              4             117              NaN
...                  ...            ...             ...              ...
2324  666411507551481857            339             459              NaN
2325  666407126856765440             44             113              NaN
2326  666396247373291520             92             172              NaN
2327  666373753744588802            100             194              NaN
2328  666362758909284353            595             804              NaN
2329  666353288456101888             77             229              NaN
2330  666345417576210432            146             307              NaN
2331  666337882303524864             96             204              NaN
2332  666293911632134144            368             522              NaN
2333  666287406224695296             71             152              NaN
2334  666273097616637952             82             184              NaN
2335  666268910803644416             37             108              NaN
2336  666104133288665088           6871           14765              NaN
2337  666102155909144576             16              81              NaN
2338  666099513787052032             73             164              NaN
2339  666094000022159362             79             169              NaN
2340  666082916733198337             47             121              NaN
2341  666073100786774016            174             335              NaN
2342  666071193221509120             67             154              NaN
2343  666063827256086533            232             496              NaN
2344  666058600524156928             61             115              NaN
2345  666057090499244032            146             304              NaN
2346  666055525042405380            261             448              NaN
2347  6

In [9]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [10]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [11]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2354 entries, 0 to 2353
Data columns (total 4 columns):
id                  2354 non-null int64
retweet_count       2354 non-null int64
favorite_count      2354 non-null int64
retweeted_status    179 non-null object
dtypes: int64(3), object(1)
memory usage: 92.0+ KB


In [12]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
pd.set_option('display.max_colwidth', -1)

twitter_archive[twitter_archive.rating_denominator == 0]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640  8.352460e+17           26259576.0            

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                                                 source  \
313  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                   text  \
313  @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
313 NaN                  NaN                        NaN                         

    expanded_urls  rating_numerator  rating_denominator  name doggo floofer  \
313  NaN           960               0                   None  None  None     

    pupper puppo  
313  None   None

In [14]:
# List all tweets which has rating denominator is not equal to 10
twitter_archive[twitter_archive.rating_denominator != 10]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640  8.352460e+17           2.625958e+07          
342   832088576586297345  8.320875e+17           3.058208e+07          
433   820690176645140481 NaN                    NaN                    
516   810984652412424192 NaN                    NaN                    
784   775096608509886464 NaN                    NaN                    
902   758467244762497024 NaN                    NaN                    
1068  740373189193256964 NaN                    NaN                    
1120  731156023742988288 NaN                    NaN                    
1165  722974582966214656 NaN                    NaN                    
1202  716439118184652801 NaN                    NaN                    
1228  713900603437621249 NaN                    NaN                    
1254  710658690886586372 NaN                    NaN                    
1274  709198395643068416 NaN                    NaN                    
1351  704054845121142784 NaN                    NaN                    
1433  697463031882764288 NaN                    NaN                    
1598  686035780142297088  6.860340e+17           4.196984e+09          
1634  684225744407494656  6.842229e+17           4.196984e+09          
1635  684222868335505415 NaN                    NaN                    
1662  682962037429899265 NaN                    NaN                    
1663  682808988178739200  6.827884e+17           4.196984e+09          
1779  677716515794329600 NaN                    NaN                    
1843  675853064436391936 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                                                  source  \
313   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
342   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
433   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
516   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
784   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
902   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1228  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1254  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1274  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1433  <a href="http://twitter.com/download/iphone" rel="nofollow">

In [15]:
# Inspect all tweets which has rating numerator is 9 and denominator is 11 to look for possible duplicate 
twitter_archive[(twitter_archive.rating_numerator == 9) & (twitter_archive.rating_denominator == 11) ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
784   775096608509886464 NaN                    NaN                    
1068  740373189193256964 NaN                    NaN                    

                      timestamp  \
784   2016-09-11 22:20:06 +0000   
1068  2016-06-08 02:41:38 +0000   

                                                                                  source  \
784   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
784   RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   

      retweeted_status_id  retweeted_status_user_id  \
784   7.403732e+17         4.196984e+09               
1068 NaN                  NaN                         

     retweeted_status_timestamp  \
784   2016-06-08 02:41:38 +0000   
1068  NaN                         

                                                                                                                                                                                                                                                        expanded_urls  \
784   https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1   
1068  https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
784   9                 11                  None  None  None    None   None  
1068  9                 11                  None  None  None    None   None

In [16]:
# Check for duplicated tweets by looking for 'RT @' in the text column
twitter_archive[twitter_archive['text'].str.contains("RT @")]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257 NaN                    NaN                    
32    886054160059072513 NaN                    NaN                    
36    885311592912609280 NaN                    NaN                    
68    879130579576475649 NaN                    NaN                    
73    878404777348136964 NaN                    NaN                    
74    878316110768087041 NaN                    NaN                    
78    877611172832227328 NaN                    NaN                    
91    874434818259525634 NaN                    NaN                    
95    873697596434513921 NaN                    NaN                    
97    873337748698140672 NaN                    NaN                    
101   872668790621863937 NaN                    NaN                    
109   871166179821445120 NaN                    NaN                    
118   869988702071779329 NaN                    NaN                    
124   868639477480148993 NaN                    NaN                    
130   867072653475098625 NaN                    NaN                    
132   866816280283807744 NaN                    NaN                    
137   866094527597207552 NaN                    NaN                    
146   863471782782697472 NaN                    NaN                    
155   861769973181624320 NaN                    NaN                    
159   860981674716409858 NaN                    NaN                    
160   860924035999428608 NaN                    NaN                    
165   860177593139703809 NaN                    NaN                    
171   858860390427611136 NaN                    NaN                    
180   857062103051644929 NaN                    NaN                    
182   856602993587888130 NaN                    NaN                    
185   856330835276025856 NaN                    NaN                    
194   855245323840757760 NaN                    NaN                    
195   855138241867124737 NaN                    NaN                    
204   852936405516943360 NaN                    NaN                    
211   851953902622658560 NaN                    NaN                    
...                  ...  ..                     ..                    
784   775096608509886464 NaN                    NaN                    
794   773336787167145985 NaN                    NaN                    
800   772615324260794368 NaN                    NaN                    
811   771171053431250945 NaN                    NaN                    
815   771004394259247104 NaN                    NaN                    
818   770743923962707968 NaN                    NaN                    
822   770093767776997377 NaN                    NaN                    
826   769335591808995329 NaN                    NaN                    
829   768909767477751808 NaN                    NaN                    
833   768554158521745409 NaN                    NaN                    
841   766864461642756096 NaN                    NaN                    
847   766078092750233600 NaN                    NaN                    
860   763167063695355904 NaN                    NaN                    
868   761750502866649088 NaN                    NaN                    
872   761371037149827077 NaN                    NaN                    
885   760153949710192640 NaN                    NaN                    
890   759566828574212096 NaN                    NaN                    
895   759159934323924993 NaN                    NaN                    
908   757729163776290825 NaN                    NaN                    
911   757597904299253760 NaN                    NaN                    
926   754874841593970688 NaN                    NaN                    
937   753298634498793472 NaN                    NaN                    
943   752701944171524096 NaN                    NaN                    
949   752309394570878976 NaN                    NaN     

In [17]:
twitter_archive[twitter_archive['text'].str.contains("RT @")]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257 NaN                    NaN                    
32    886054160059072513 NaN                    NaN                    
36    885311592912609280 NaN                    NaN                    
68    879130579576475649 NaN                    NaN                    
73    878404777348136964 NaN                    NaN                    
74    878316110768087041 NaN                    NaN                    
78    877611172832227328 NaN                    NaN                    
91    874434818259525634 NaN                    NaN                    
95    873697596434513921 NaN                    NaN                    
97    873337748698140672 NaN                    NaN                    
101   872668790621863937 NaN                    NaN                    
109   871166179821445120 NaN                    NaN                    
118   869988702071779329 NaN                    NaN                    
124   868639477480148993 NaN                    NaN                    
130   867072653475098625 NaN                    NaN                    
132   866816280283807744 NaN                    NaN                    
137   866094527597207552 NaN                    NaN                    
146   863471782782697472 NaN                    NaN                    
155   861769973181624320 NaN                    NaN                    
159   860981674716409858 NaN                    NaN                    
160   860924035999428608 NaN                    NaN                    
165   860177593139703809 NaN                    NaN                    
171   858860390427611136 NaN                    NaN                    
180   857062103051644929 NaN                    NaN                    
182   856602993587888130 NaN                    NaN                    
185   856330835276025856 NaN                    NaN                    
194   855245323840757760 NaN                    NaN                    
195   855138241867124737 NaN                    NaN                    
204   852936405516943360 NaN                    NaN                    
211   851953902622658560 NaN                    NaN                    
...                  ...  ..                     ..                    
784   775096608509886464 NaN                    NaN                    
794   773336787167145985 NaN                    NaN                    
800   772615324260794368 NaN                    NaN                    
811   771171053431250945 NaN                    NaN                    
815   771004394259247104 NaN                    NaN                    
818   770743923962707968 NaN                    NaN                    
822   770093767776997377 NaN                    NaN                    
826   769335591808995329 NaN                    NaN                    
829   768909767477751808 NaN                    NaN                    
833   768554158521745409 NaN                    NaN                    
841   766864461642756096 NaN                    NaN                    
847   766078092750233600 NaN                    NaN                    
860   763167063695355904 NaN                    NaN                    
868   761750502866649088 NaN                    NaN                    
872   761371037149827077 NaN                    NaN                    
885   760153949710192640 NaN                    NaN                    
890   759566828574212096 NaN                    NaN                    
895   759159934323924993 NaN                    NaN                    
908   757729163776290825 NaN                    NaN                    
911   757597904299253760 NaN                    NaN                    
926   754874841593970688 NaN                    NaN                    
937   753298634498793472 NaN                    NaN                    
943   752701944171524096 NaN                    NaN                    
949   752309394570878976 NaN                    NaN     

In [18]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

In [19]:
tweet_json.describe()

id  retweet_count  favorite_count
count  2.354000e+03  2354.000000    2354.000000   
mean   7.426978e+17  3164.797366    8080.968564   
std    6.852812e+16  5284.770364    11814.771334  
min    6.660209e+17  0.000000       0.000000      
25%    6.783975e+17  624.500000     1415.000000   
50%    7.194596e+17  1473.500000    3603.500000   
75%    7.993058e+17  3652.000000    10122.250000  
max    8.924206e+17  79515.000000   132810.000000

In [20]:
twitter_archive.rating_denominator.value_counts()

10     2333
11     3   
50     3   
80     2   
20     2   
2      1   
16     1   
40     1   
70     1   
15     1   
90     1   
110    1   
120    1   
130    1   
150    1   
170    1   
7      1   
0      1   
Name: rating_denominator, dtype: int64

In [21]:
twitter_archive.name.value_counts()

None           745
a              55 
Charlie        12 
Cooper         11 
Oliver         11 
Lucy           11 
Lola           10 
Tucker         10 
Penny          10 
Bo             9  
Winston        9  
Sadie          8  
the            8  
Toby           7  
Buddy          7  
Daisy          7  
an             7  
Bailey         7  
Jax            6  
Bella          6  
Stanley        6  
Scout          6  
Jack           6  
Leo            6  
Milo           6  
Rusty          6  
Koda           6  
Oscar          6  
Dave           6  
George         5  
              ..  
Diogi          1  
Kanu           1  
Scott          1  
Pip            1  
Yoda           1  
Tedrick        1  
Willie         1  
Strudel        1  
Banditt        1  
Al             1  
Crawford       1  
Bloo           1  
River          1  
Dallas         1  
Spanky         1  
Corey          1  
Timmy          1  
Cecil          1  
Comet          1  
infuriating    1  
Noah           1  
Toffee      

In [22]:
twitter_archive[twitter_archive.name == 'a']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706 NaN                    NaN                    
649   792913359805018113 NaN                    NaN                    
801   772581559778025472 NaN                    NaN                    
1002  747885874273214464 NaN                    NaN                    
1004  747816857231626240 NaN                    NaN                    
1017  746872823977771008 NaN                    NaN                    
1049  743222593470234624 NaN                    NaN                    
1193  717537687239008257 NaN                    NaN                    
1207  715733265223708672 NaN                    NaN                    
1340  704859558691414016 NaN                    NaN                    
1351  704054845121142784 NaN                    NaN                    
1361  703079050210877440 NaN                    NaN                    
1368  702539513671897089 NaN                    NaN                    
1382  700864154249383937 NaN                    NaN                    
1499  692187005137076224 NaN                    NaN                    
1737  679530280114372609 NaN                    NaN                    
1785  677644091929329666 NaN                    NaN                    
1853  675706639471788032 NaN                    NaN                    
1854  675534494439489536 NaN                    NaN                    
1877  675109292475830276 NaN                    NaN                    
1878  675047298674663426 NaN                    NaN                    
1923  674082852460433408 NaN                    NaN                    
1941  673715861853720576 NaN                    NaN                    
1955  673636718965334016 NaN                    NaN                    
1994  672604026190569472 NaN                    NaN                    
2034  671743150407421952 NaN                    NaN                    
2066  671147085991960577 NaN                    NaN                    
2116  670427002554466305 NaN                    NaN                    
2125  670361874861563904 NaN                    NaN                    
2128  670303360680108032 NaN                    NaN                    
2146  669923323644657664 NaN                    NaN                    
2153  669661792646373376 NaN                    NaN                    
2161  669564461267722241 NaN                    NaN                    
2191  668955713004314625 NaN                    NaN                    
2198  668815180734689280 NaN                    NaN                    
2211  668614819948453888 NaN                    NaN                    
2218  668507509523615744 NaN                    NaN                    
2222  668466899341221888 NaN                    NaN                    
2235  668171859951755264 NaN                    NaN                    
2249  667861340749471744 NaN                    NaN                    
2255  667773195014021121 NaN                    NaN                    
2264  667538891197542400 NaN                    NaN                    
2273  667470559035432960 NaN                    NaN                    
2287  667177989038297088 NaN                    NaN                    
2304  666983947667116034 NaN                    NaN                    
2311  666781792255496192 NaN                    NaN                    
2314  666701168228331520 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2350  666050758794694657 NaN                    NaN                    
2352  666044226329800704 NaN                    NaN                    
2353  666033412701032449 NaN                    NaN                    
2354  666029285002620928 NaN                    NaN     

In [23]:
twitter_archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
420     2  
0       2  
15      2  
75      2  
80      1  
20      1  
24      1  
26      1  
44      1  
50      1  
60      1  
165     1  
84      1  
88      1  
144     1  
182     1  
143     1  
666     1  
960     1  
1776    1  
17      1  
27      1  
45      1  
99      1  
121     1  
204     1  
Name: rating_numerator, dtype: int64

#### Quality
##### `twitterArchive_df` table
- Change timestamp column from object to date time format

- Remove 181 tweets (duplicate rating because it is retweet)

- Drop in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id ,expanded_urls,  retweeted_status_user_id and retweeted_status_timestamp because their records are useless due to the lack of information (NAN valuesare not usefull for analysis)

- Fix tweet ID 835246439529840640 that has rating_denominator 0 (not correct mathmetically). it should be 13/10

- Remove tweet ID 740373189193256964 because the rating is not correct (refer back to 'text' column 9/11 is date not rate)

- Remove tweet ID 810984652412424192 because tweet dose not contain rating (user mentioned 24 hours/ 7 days).
- Remove tweet id 832088576586297345 since it contains no rating 


- Fix the rating for tweet id 722974582966214656 (rating 13/10 not 4/20)

- Fix the rating for tweet id 666287406224695296 (rating 9/10 not 3/1)

- Fix the rating for tweet id 682962037429899265 (rating 10/10 not 7/11)

- Fix the rating for tweet id 716439118184652801 (rating 11/10 not 50/50)

- Calculate the rating and store in 'rating' column in twitter_archive



#### Tidiness
- Condensing Dog Type columns
- The dog stages have values as columns, instead of one column filled with their values.
- Condensing dog breed predictions


## Clean

In [24]:
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean = image_predictions.copy()
tweet_json_clean = tweet_json.copy()

#### `twitter_archive`: Incorrect  Column Type 

##### Define
Change timestamp column from object to date time format

##### Code

In [25]:
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean['timestamp'])

##### Test

In [26]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](1

#### `twitter_archive`: Remove Retweets and Tweets which does not include image

##### Define
Remove 181 tweets (duplicate rating because it is retweet)

##### Code

In [27]:
# removing retweets
twitter_archive_clean = twitter_archive_clean[pd.isnull(twitter_archive_clean['retweeted_status_id'])]
twitter_archive_clean.shape[0]

2175

In [28]:
twitter_archive_clean.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis = 1, inplace = True)


##### Test

In [29]:
twitter_archive_clean.shape[0]

2175

In [30]:
# removing the tweets without images
list(twitter_archive_clean)


['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

#### `twitter_archive`: Data quality issues

##### Define
Fix tweet ID 835246439529840640 that has rating_denominator 0 (not correct mathmetically). it should be 13/10

##### Code

In [31]:
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 835246439529840640, 'rating_numerator'] = 13
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 835246439529840640, 'rating_denominator'] = 10


##### Test

In [32]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 835246439529840640 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640  8.352460e+17           26259576.0            

              timestamp  \
313 2017-02-24 21:54:03   

                                                                                 source  \
313  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                   text  \
313  @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   

    expanded_urls  rating_numerator  rating_denominator  name doggo floofer  \
313  NaN           13                10                  None  None  None     

    pupper puppo  
313  None   None

##### Define
- Remove tweet ID 740373189193256964 because the rating is not correct (refer back to 'text' column 9/11 is date not rate)
- Remove tweet ID 810984652412424192 because tweet dose not contain rating (user mentioned 24 hours/ 7 days). it should be 13/10
- Remove tweet id 832088576586297345 since it contains no rating 


##### Code

In [33]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.tweet_id != 740373189193256964]
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.tweet_id != 810984652412424192]
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.tweet_id != 832088576586297345]


##### Test

In [34]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 740373189193256964 ]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [35]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 810984652412424192 ]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [36]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 832088576586297345 ]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

##### Define
- Fix the rating for tweet id 722974582966214656 (rating 13/10 not 4/20)
- Fix the rating for tweet id 666287406224695296 (rating 9/10 not 3/1)
- Fix the rating for tweet id 682962037429899265 (rating 10/10 not 7/11)
- Fix the rating for tweet id 716439118184652801 (rating 11/10 not 50/50)


##### Code

In [37]:

twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 722974582966214656, 'rating_numerator'] = 13
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 722974582966214656, 'rating_denominator'] = 10

twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 666287406224695296, 'rating_numerator'] = 9
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 666287406224695296, 'rating_denominator'] = 10


twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 682962037429899265, 'rating_numerator'] = 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 682962037429899265, 'rating_denominator'] = 10

twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 716439118184652801, 'rating_numerator'] = 11
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 716439118184652801, 'rating_denominator'] = 10


##### Test

In [38]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 722974582966214656 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1165  722974582966214656 NaN                    NaN                    

               timestamp  \
1165 2016-04-21 02:25:47   

                                                                                  source  \
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                  text  \
1165  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   

                                                        expanded_urls  \
1165  https://twitter.com/dog_rates/status/722974582966214656/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
1165  13                10                  None  None  None    None   None

In [39]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 666287406224695296 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2335  666287406224695296 NaN                    NaN                    

               timestamp  \
2335 2015-11-16 16:11:11   

                                                                                  source  \
2335  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                          text  \
2335  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

                                                        expanded_urls  \
2335  https://twitter.com/dog_rates/status/666287406224695296/photo/1   

      rating_numerator  rating_denominator name doggo floofer pupper puppo  
2335  9                 10                  an   None  None    None   None

In [40]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 682962037429899265 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1662  682962037429899265 NaN                    NaN                    

               timestamp  \
1662 2016-01-01 16:30:13   

                                                                                  source  \
1662  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                             text  \
1662  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   

                                                        expanded_urls  \
1662  https://twitter.com/dog_rates/status/682962037429899265/photo/1   

      rating_numerator  rating_denominator    name doggo floofer pupper puppo  
1662  10                10                  Darrel  None  None    None   None

In [41]:
twitter_archive_clean[twitter_archive_clean.tweet_id == 716439118184652801 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1202  716439118184652801 NaN                    NaN                    

               timestamp  \
1202 2016-04-03 01:36:11   

                                                                                  source  \
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                       text  \
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   

                                                        expanded_urls  \
1202  https://twitter.com/dog_rates/status/716439118184652801/photo/1   

      rating_numerator  rating_denominator      name doggo floofer pupper  \
1202  11                10                  Bluebert  None  None    None    

     puppo  
1202  None

##### Define
- Calculate the rating and store in 'rating' column in twitter_archive

##### Code

In [42]:
twitter_archive_clean['rating']= twitter_archive_clean['rating_numerator'] / twitter_archive_clean['rating_denominator']

In [43]:
twitter_archive_clean.drop(['rating_numerator', 'rating_denominator'], axis = 1, inplace = True)


##### Test

In [44]:
twitter_archive_clean.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1818  676593408224403456 NaN                    NaN                    
1777  677918531514703872 NaN                    NaN                    

               timestamp  \
1818 2015-12-15 02:43:33   
1777 2015-12-18 18:29:07   

                                                                                  source  \
1818  <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                      
1777  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
1818  This pupper loves leaves. 11/10 for committed leaf lover https://t.co/APvLqbEhkF                                                               
1777  "Everything looks pretty good in there. Make sure to brush your gums. Been flossing? How's school going?" Both 10/10 https://t.co/lWL2IMJqLR   

                                                        expanded_urls  name  \
1818  https://vine.co/v/eEQQaPFbgOY                                    None   
1777  https://twitter.com/dog_rates/status/677918531514703872/photo/1  None   

     doggo floofer  pupper puppo  rating  
1818  None  None    pupper  None  1.1     
1777  None  None    None    None  1.0

#### `twitter_archive`: Tidiness issues

##### Define
- Condensing Dog Type columns

##### Code

In [45]:
dog_type = []

x = ['pupper', 'puppo', 'doggo', 'floof']
y = ['pupper', 'puppo', 'doggo', 'floof']

for row in twitter_archive_clean['text']:
    row = row.lower()
    for word in x:
        if word in str(row):
            dog_type.append(y[x.index(word)])
            break
    else:
        dog_type.append('None')

twitter_archive_clean['dog_type'] = dog_type

In [46]:
# removing extra columns
twitter_archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1, inplace=True)


##### Test

In [47]:
twitter_archive_clean['dog_type'].value_counts()

None      1760
pupper    260 
doggo     83  
floof     36  
puppo     33  
Name: dog_type, dtype: int64

In [48]:
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'name',
 'rating',
 'dog_type']

#### `image_predictions`: Tidiness issues

##### Define
- Condensing dog breed predictions

##### Code

In [49]:
# Store all the data in the breed and conf variables
breed = []
conf= []

# Create function where you will loop through the image predictions data frame
# and check p1_dog to p3_dog . If all of them are true then it will take the highest p1_dog and their conf
# if none of them are true, then it will assign Unidentifiable and 0 conf
def breed_conf(row):
    if row['p1_dog']:
        breed.append(row['p1'])
        conf.append(row['p1_conf'])
    elif row['p2_dog']:
        breed.append(row['p2'])
        conf.append(row['p2_conf'])
    elif row['p3_dog']:
        breed.append(row['p3'])
        conf.append(row['p3_conf'])
    else:
        breed.append('Unidentifiable')
        conf.append(0)

# use apply to apply function on the data frame
image_predictions_clean.apply(breed_conf, axis = 1)

# assign the results to data frame
image_predictions_clean['breed'] = breed
image_predictions_clean['confidence'] = conf


In [50]:
#removing the processed columns
image_predictions_clean.drop(['p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog',], axis=1, inplace=True)


##### Test

In [51]:
image_predictions_clean.sample(4)

tweet_id                                          jpg_url  \
1388  766313316352462849  https://pbs.twimg.com/media/CqJ95SRWgAATPK_.jpg   
45    666786068205871104  https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
1409  770414278348247044  https://pbs.twimg.com/media/CrEPsfWXEAAKvem.jpg   
1243  747219827526344708  https://pbs.twimg.com/media/Cl6odlVWQAIy5uk.jpg   

      img_num              breed  confidence  
1388  1        toy_poodle         0.966896    
45    1        Unidentifiable     0.000000    
1409  1        golden_retriever   0.053570    
1243  2        Shetland_sheepdog  0.548018

In [52]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 5 columns):
tweet_id      2075 non-null int64
jpg_url       2075 non-null object
img_num       2075 non-null int64
breed         2075 non-null object
confidence    2075 non-null float64
dtypes: float64(1), int64(2), object(2)
memory usage: 81.1+ KB


#### Combine  all the three resource in one master data frame:

In [53]:
twitter_archive = twitter_archive_clean.copy()
image_predictions = image_predictions_clean.copy()
tweet_json = tweet_json_clean.copy()

#### Start migrate all the 3 collected data source into one master file

In [54]:
# Set the index for image prediction and tweet json
#imagePredictions_df.set_index('tweet_id', inplace = True)
#tweet_json.set_index('id', inplace = True)

master_df = pd.merge(left=twitter_archive, right=image_predictions, left_index=True, right_index=True, how='left')
master_df = pd.merge(left=master_df, right=tweet_json, left_index=True, right_index=True, how='left')


In [55]:
master_df

tweet_id_x  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
30    886267009285017600  8.862664e+17           2.281182e+09          
...                  ...           ...                    ...          
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2328  666396247373291520 NaN                    NaN                    
2329  666373753744588802 NaN                    NaN                    
2330  666362758909284353 NaN                    NaN                    
2331  666353288456101888 NaN                    NaN                    
2332  666345417576210432 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2336  666273097616637952 NaN                    NaN                    
2337  666268910803644416 NaN                    NaN                    
2338  666104133288665088 NaN                    NaN                    
2339  666102155909144576 NaN                    NaN                    
2340  666099513787052032 NaN                    NaN                    
2341  666094000022159362 NaN                    NaN                    
2342  666082916733198337 NaN                    NaN                    
2343  666073100786774016 NaN                    NaN                    
2344  666071193221509120 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2349  666051853826850816 NaN                    NaN   

#### `master_df`: Polishing the data frame

##### Define
- Remove in_reply_to_status_id, in_reply_to_user_id, tweet_id_y and retweeted_status

##### Code

In [56]:
master_df.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'tweet_id_y', 'retweeted_status', 'id'], axis = 1, inplace = True)


##### Test

In [57]:
master_df.to_csv('master_df.csv', encoding = 'utf-8')
master_df.head()

tweet_id_x           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

       name  rating dog_type                                          jpg_url  \
0  Phineas   1.3     None     https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  Tilly     1.3     None     https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  Archie    1.2     None     https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  Darla     1.3     None     https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  Franklin  1.2     None     https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                   breed  confidence  retweet_count  favorite_count  
0  1.0      Welsh_springer_spaniel  0.465074    8853.0         39467.0         
1  1.0      redbone                 0.506826    6514.0         33819.0         
2  1.0      German_shepherd         0.596461    4328.0         25461.0         
3  1.0      Rhodesian_ridgeback     0.408143    8964.0         42908.0         
4  1.0      miniature_pinscher      0.560311    9774.0         41048.0